### PR TITLE
Added Private Messaging Feature Using Socket.io and Sequelize

### PR DESCRIPTION
**Pull request by javidangarcia**
- Added Message model and endpoints to store new messages and retrieve past messages.
- Added /friends endpoint so that the user can select what friends to chat with.
- Used the rooms functionality of socket.io to implement private messaging with many users.
- Used socket.io to listen for sent messages and broadcast them in real time to the other user in the room.

https://github.com/javidangarcia/stock-zone/assets/89328228/89d5b155-d908-4e47-a271-517439fe0821

https://github.com/javidangarcia/stock-zone/assets/89328228/1e560b08-4c2c-4172-932d-7a3ce2458175

### CODE REVIEW

File: frontend/src/components/ChatRoom/ChatRoom.jsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        setShowChat(true);
+    };
+
+    return (
+        <div className="chat">
+            <div className="friends-list">
+                <h3>Friends</h3>
+                <Card className="friends-card">
+                    <ListGroup variant="flush">
+                        {friends.map((friend) => (
```
Comment (BriptimusPrimus) [inline frontend/src/components/ChatRoom/ChatRoom.jsx:62]: the object friends have a couple of users that are connected as friends.
A better name for that entity would be "connection" or "friendship" or "pair".

File: frontend/src/components/Chat/Chat.jsx [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                }
+
+                if (response.status === 500) {
+                    setErrorMessage(response.data.error);
+                }
+            } catch (error) {
+                setErrorMessage(`${error.message}: Please try again later.`);
+            }
+            const response = await axios.get(
+                `${import.meta.env.VITE_HOST}/messages/${friend.user2.id}`,
```
Comment (BriptimusPrimus) [inline frontend/src/components/Chat/Chat.jsx:32]: Why is this call repeated?

File: frontend/src/components/Chat/Chat.jsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            setErrorMessage(error);
+        }
+    };
+
+    useEffect(() => {
+        socket.on("receive_message", (data) => {
+            setMessageList((list) => [...list, data]);
+        });
+
+        return () => socket.removeListener("receive_message");
```
Comment (BriptimusPrimus) [inline frontend/src/components/Chat/Chat.jsx:79]: 👍

File: frontend/src/components/Chat/Chat.jsx [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    return (
+        <div className="chat-section">
+            <ScrollToBottom className="chat-window">
+                {messageList.map((messageContent) => {
+                    return (
+                        <div
+                            key={messageContent.id}
+                            className="message"
+                            id={
```
Comment (BriptimusPrimus) [inline frontend/src/components/Chat/Chat.jsx:90]: An ID should be unique in the HTML document. This is creating N number of divs with id="you" and M number of divs with id="other"

**Comment by BriptimusPrimus** [comment]
Good job!

File: frontend/src/components/Chat/Chat.jsx (hunk 2 above)
Comment (javidangarcia) [reply frontend/src/components/Chat/Chat.jsx:32]: That call is to get the chat history with your friend. The call in ChatRoom just gets a list of all the friends you have.

**Pull request merged**